tests/test_client_logger.cpp

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (C) Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "mock_server_reader_writer.h"
+#include "stub_logger.h"
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+#include <multipass/logging/client_logger.h>
+#include <multipass/logging/level.h>
+#include <multipass/logging/multiplexing_logger.h>
+
+namespace mpl = multipass::logging;
+namespace mpt = multipass::test;
+
+struct StubReply
+{
+    template <typename... Args>
+    void set_log_line(const std::string& msg)
```

Comment:
This template here seems suspicious, perhaps a leftover?

Reply:
Nope. `ClientLogger.log()` calls `reply.set_log_line()` to preserve the produced log line in the reply object. Then, this object is passed to the `server->Write()` function. The mock `EXPECT_CALL(..., Write)` verifies that `stored_msg` contains the expected log message.

Reply:
Done